### PR TITLE
perf: keep jobs in waiting list when queue is paused

### PR DESCRIPTION
**Pull request by roggervalf**
This PR removes the legacy code regarding the old paused queue functionality, which was based on renaming the wait list key to paused. As paused keys are now handled as a queue state, we do not need the old mechanism anymore and so the code can be removed and simplified. We need some migration steps for old queues that may be in paused status at the time of the upgrade.

### CODE REVIEW

File: docs/gitbook/guide/flows/fail-parent.md [hunk 1]
```diff
@@ -16,13 +16,13 @@ const originalTree = await flow.add({
       name,
       data: { idx: 0, foo: 'bar' },
       queueName: 'childrenQueueName',
-      opts: { failParentOnFailure: true },
+      opts: { onChildFailure: 'fail' },
```
Comment (manast) [inline docs/gitbook/guide/flows/fail-parent.md:19]: This is a nice change.

File: python/bullmq/scripts.py [hunk 2]
```diff
@@ -451,6 +462,15 @@ async def retryJobs(self, state: str, count: int, timestamp: int):
         result = await self.commands["moveJobsToWait"](keys=keys, args=args)
         return result
 
+    async def repairDeprecatedPausedKey(self, maxCount: int):
```
Comment (manast) [inline python/bullmq/scripts.py:465]: We need proper documentation on how and when a call to this method is needed.

File: src/classes/job.ts [hunk 3]
```diff
@@ -39,10 +39,8 @@ const logger = debuglog('bull');
 
 const optsDecodeMap = {
   de: 'debounce',
-  fpof: 'failParentOnFailure',
```
Comment (manast) [inline src/classes/job.ts:42]: Won't we need to keep this old mappings to not cause a data breaking change?

Comment (manast) [reply src/classes/job.ts:42]: Maybe we should introduce a "migration" mechanism here. Something like migration steps for going between versions that are run if required in Lua scripts for atomicity. But we would need to keep a version number in the meta key, adding complexity.

File: src/classes/job.ts [hunk 4]
```diff
@@ -1210,27 +1208,6 @@ export class Job<
       throw new Error(`Delay and repeat options could not be used together`);
     }
 
-    if (this.opts.removeDependencyOnFailure && this.opts.failParentOnFailure) {
```
Comment (manast) [inline src/classes/job.ts:1213]: nice to be able to remove this

File: src/classes/queue.ts [hunk 5]
```diff
@@ -455,6 +454,13 @@ export class Queue<
     return this.scripts.drain(delayed);
   }
 
+  /**
+   * Remove legacy markers before v5
+   */
+  removeLegacyMarkers(): Promise<void> {
```
Comment (manast) [inline src/classes/queue.ts:460]: Documentation about this is also needed. Most importantly practical information in how to apply something like this in the context of a production deployment.

File: src/classes/queue.ts [hunk 6]
```diff
@@ -580,4 +586,18 @@ export class Queue<
     const client = await this.client;
     return client.del(this.toKey('priority'));
   }
+
+  /**
+   * Repair deprecated paused key
+   *
+   * @param maxCount - Max quantity of jobs to be moved to wait per iteration.
+   */
+  async repairDeprecatedPausedKey(maxCount: number = 1000): Promise<void> {
```
Comment (manast) [inline src/classes/queue.ts:595]: Should "repair" be the proper name or actually something like "migrate" would be more accurate?

**Review by manast — COMMENTED** [review]
Some initial comments, more will come.

File: src/classes/job.ts (hunk 3 above)
Comment (roggervalf) [reply src/classes/job.ts]: we can keep this mappings that will be only used to format options of jobs with these old values, but no migration is needed as we are evaluation old and new option in lua scripts, new jobs will use new option

File: src/classes/queue.ts (hunk 6 above)
Comment (roggervalf) [reply src/classes/queue.ts:595]: I can change it, let me do it

File: src/classes/queue.ts (hunk 5 above)
Comment (roggervalf) [reply src/classes/queue.ts]: I added a me migrations section for it

File: docs/gitbook/guide/migrations/v6.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+# Migration to v6
+
+Make sure to pass **skipMigrationsExecution** option in any of our instances as false in order to execute all necessary changes when coming from an older version
+
+## Migration of deprecated paused key
+
+If you have paused queues after upgrading to this version. These jobs will be moved to wait state when initializing any of our instances (Worker, Queue, QueueEvents or FlowProducer).
+
+Paused key is not longer needed as this state is already represented by queue meta key. It also improve the process of pausing or resuming a queue as we don't need to rename any key.
```
Comment (manast) [inline docs/gitbook/guide/migrations/v6.md:13]: by meta key -> inside the meta key

Comment (manast) [reply docs/gitbook/guide/migrations/v6.md:13]: "It also improves..."

File: docs/gitbook/guide/migrations/v6.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+## Remove legacy markers
```
Comment (manast) [inline docs/gitbook/guide/migrations/v6.md:15]: I think we are going to need more detailed information on how to safely perform the migration.

File: src/classes/queue-base.ts [hunk 9]
```diff
@@ -80,9 +84,40 @@ export class QueueBase extends EventEmitter implements MinimalQueue {
...
+          if (hasPendingMigrations) {
+            if (
+              this.opts.skipMigrationsExecution ||
+              typeof this.opts.skipMigrationsExecution === 'undefined'
+            ) {
+              console.error(
+                'Queue has pending migrations. See https://docs.bullmq.io/guide/migrations',
+              );
+            } else {
+              return runMigrations(client, {
```
Comment (manast) [inline src/classes/queue-base.ts:107]: I do not think we should silently run migrations here, this just opens a window to shoot ourselves in the foot.

File: src/classes/queue.ts [hunk 10]
```diff
@@ -709,4 +709,16 @@ export class Queue<
     const client = await this.client;
     return client.del(this.toKey('priority'));
   }
+
+  /**
+   * Migrate deprecated paused key
+   *
+   * @param maxCount - Max quantity of jobs to be moved to wait per iteration.
+   */
+  async migrateDeprecatedPausedKey(maxCount = 1000): Promise<void> {
```
Comment (manast) [inline src/classes/queue.ts:718]: should be moved to a migration step.

File: src/commands/migrateDeprecatedPausedKey-2.lua [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  Move paused job ids to wait state to repair these states
+
+  Input:
+    KEYS[1] paused key
+    KEYS[2] wait key
+
+    ARGV[1] count
+]]
+
+local maxCount = tonumber(ARGV[1])
```
Comment (manast) [inline src/commands/migrateDeprecatedPausedKey-2.lua:11]: I do not understand this script. It should not be possible to have both wait and paused keys at the same time, either it is wait or paused but not both, so basically just renaming the key from paused to wait should be enough as a migration step + setting the queue status to paused.

**Review by manast — APPROVED** [review]
Just a small comment in the python version.

File: python/bullmq/scripts.py [hunk 12]
```diff
@@ -465,6 +463,15 @@ async def retryJobs(self, state: str, count: int, timestamp: int):
         result = await self.commands["moveJobsToWait"](keys=keys, args=args)
         return result
 
+    async def migrateDeprecatedPausedKey(self, maxCount: int):
```
[the remark this answers is not in the document]
Comment (roggervalf) [reply python/bullmq/scripts.py]: I'll add this logic in a following pr

**Pull request merged**